soroban-sdk-tools-macro/src/error.rs

```diff
@@ -1,30 +1,678 @@
+                ));
+            }
+            if code > root_max {
+                return Err(Error::new(
+                    variant.span(),
+                    format!(
+                        "Root mode error variant index {} exceeds maximum of {} ({}-bit limit). \
+                        Consider splitting into multiple error enums or reducing variant count.",
+                        code, root_max, ROOT_BITS
+                    )
+                ));
+            }
+        }
+
+        // Collision Check
+        if let Some(existing_variant) = used_codes.get(&code) {
+             let msg = format!(
+                "Error code collision: Variant '{}' (code {}) conflicts with '{}'. \
+                Please manually assign a unique discriminant (e.g., `{} = {}`) to resolve this.",
+                variant.ident, code, existing_variant, variant.ident, code + 1
+            );
+            return Err(Error::new(variant.span(), msg));
+        }
+
+        used_codes.insert(code, variant.ident.to_string());
+        codes.push(code);
+    }
+
+    Ok(codes)
+}
+
+fn collect_variant_infos<'a>(
+    data: &'a DataEnum,
+    enum_ident: &'a Ident,
+    mode: ScerrMode,
+) -> syn::Result<Vec<VariantInfo<'a>>> {
+    let codes = assign_codes(data, enum_ident, mode)?;
+    data.variants
+        .iter()
+        .zip(codes)
+        .map(|(v, code)| parse_variant_info(v, code))
+        .collect()
+}
+
+fn validate_variants(infos: &[VariantInfo], mode: ScerrMode) -> syn::Result<()> {
+    let mut errors = Vec::new();
+
+    for info in infos {
+        let is_unit = info.field_ty.is_none();
+        if info.is_transparent && info.is_fcc {
+            errors.push(Error::new(info.variant.span(), "Variant cannot have both #[transparent] and #[from_contract_client]"));
+        }
+        if info.is_from_abort && !is_unit {
+            errors.push(Error::new(info.variant.span(), "#[from_abort_error] only allowed on unit variants"));
+        }
+
+        match mode {
+            ScerrMode::Basic => {
+                if !is_unit {
+                    errors.push(Error::new(info.variant.span(), "Non-unit variants not supported in basic mode. Use #[scerr(root)]"));
+                }
+            }
+            ScerrMode::Root => {
+                if !is_unit && !info.is_transparent && !info.is_fcc {
+                    errors.push(Error::new(info.variant.span(), "Non-unit variants must have #[transparent] or #[from_contract_client]"));
+                }
+            }
+        }
+    }
+
+    // Note: We don't require a sentinel; we panic if unhandled.
+    // However, if the user WANTS to handle invoke errors, they need at least one FCC or Abort handler.
+    // If they just have transparent mappings, that's valid too (just remapping specific codes).
+
+    if let Some(combined_error) = errors.into_iter().reduce(|mut a, b| { a.combine(b); a }) {
```

Comment:
Can't we reuse the other one for storage? Perhaps move it to an `errors.rs` file?